beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -0,0 +1,307 @@
+	_, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.ProcessAttestation")
+	defer span.End()
+	f.votesLock.Lock()
+	defer f.votesLock.Unlock()
+
+	for _, index := range validatorIndices {
+		// Validator indices will grow the vote cache.
+		for index >= uint64(len(f.votes)) {
+			f.votes = append(f.votes, Vote{currentRoot: params.BeaconConfig().ZeroHash, nextRoot: params.BeaconConfig().ZeroHash})
+		}
+
+		// Newly allocated vote if the root fields are untouched.
+		newVote := f.votes[index].nextRoot == params.BeaconConfig().ZeroHash &&
+			f.votes[index].currentRoot == params.BeaconConfig().ZeroHash
+
+		// Vote gets updated if it's newly allocated or high target epoch.
+		if newVote || targetEpoch > f.votes[index].nextEpoch {
+			f.votes[index].nextEpoch = targetEpoch
+			f.votes[index].nextRoot = blockRoot
+		}
+	}
+
+	processedAttestationCount.Inc()
+}
+
+// ProcessBlock processes a new block by inserting it to the fork choice store.
+func (f *ForkChoice) ProcessBlock(
+	ctx context.Context,
+	slot types.Slot,
+	blockRoot, parentRoot [fieldparams.RootLength]byte,
+	justifiedEpoch, finalizedEpoch types.Epoch, optimistic bool,
+) error {
+	ctx, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.ProcessBlock")
+	defer span.End()
+
+	return f.store.insert(ctx, slot, blockRoot, parentRoot, justifiedEpoch, finalizedEpoch, optimistic)
+}
+
+// Prune prunes the fork choice store with the new finalized root. The store is only pruned if the input
+// root is different than the current store finalized root, and the number of the store has met prune threshold.
+func (f *ForkChoice) Prune(ctx context.Context, finalizedRoot [32]byte) error {
+	return f.store.prune(ctx, finalizedRoot)
+}
+
+// HasNode returns true if the node exists in fork choice store,
+// false else wise.
+func (f *ForkChoice) HasNode(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	_, ok := f.store.nodeByRoot[root]
+	return ok
+}
+
+// HasParent returns true if the node parent exists in fork choice store,
+// false else wise.
+func (f *ForkChoice) HasParent(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	node, ok := f.store.nodeByRoot[root]
+	if !ok || node == nil {
+		return false
+	}
+
+	return node.parent != nil
+}
+
+// IsCanonical returns true if the given root is part of the canonical chain.
+func (f *ForkChoice) IsCanonical(root [32]byte) bool {
+	f.store.nodesLock.RLock()
+	defer f.store.nodesLock.RUnlock()
+
+	node, ok := f.store.nodeByRoot[root]
+	if !ok || node == nil {
+		return false
+	}
+
+	if node.bestDescendant == nil {
+		if f.store.headNode.bestDescendant == nil {
+			return node == f.store.headNode
+		}
+		return node == f.store.headNode.bestDescendant
+	}
+	if f.store.headNode.bestDescendant == nil {
+		return node.bestDescendant == f.store.headNode
+	}
+	return node.bestDescendant == f.store.headNode.bestDescendant
```

Comment:
I need a rewind here. The reason we can't just:
`return node.bestDescendant == f.store.headNode`

Is because we don't update `s.headNode` on `insert()` ?

Reply:
correct, and we cannot update head because in insert we are not requesting the justified root. I talked to @prestonvanloon about the current interface to forkchoice which takes these functions with too many arguments and will refactor this in future PRs: we should just take a block's pointer to forkchoice's `ProcessBlock` and let forkchoice grab those arguments. This way we can immediately update head and this check here simplifies a bit